src/main/java/io/micronaut/build/testresources/TestResourcesLifecycleExtension.java

```diff
@@ -44,89 +51,98 @@ public class TestResourcesLifecycleExtension extends AbstractMavenLifecycleParti
     private static final String EXPLICIT_START_SERVICE_GOAL_NAME = "mn:" + StartTestResourcesServerMojo.NAME;
     private static final String EXPLICIT_STOP_SERVICE_GOAL_NAME = "mn:" + StopTestResourcesServerMojo.NAME;
 
+    private ExpressionEvaluator evaluator;
+
     @Override
     public void afterProjectsRead(MavenSession session) {
-        List<String> goals = session.getGoals();
-        if (goals.stream().anyMatch(EXPLICIT_START_SERVICE_GOAL_NAME::equals)) {
-            // we need to keep the server alive at the end of the build
-            session.getAllProjects().forEach(p -> {
-                Build build = p.getBuild();
-                withPlugin(build, "micronaut-maven-plugin", plugin -> {
-                    Xpp3Dom configuration = (Xpp3Dom) plugin.getConfiguration();
-                    boolean enabled = isEnabled(configuration, p);
-                    if (enabled) {
-                        if (configuration == null) {
-                            configuration = new Xpp3Dom("configuration");
-                            plugin.setConfiguration(configuration);
-                        }
+        session.getAllProjects().forEach(p -> {
+            Build build = p.getBuild();
+            withPlugin(build, THIS_PLUGIN, plugin -> {
+                this.evaluator = getEvaluator(session, plugin);
```

Comment:
I'm personally not fond of stateful fields like this: it's hard to read. I was looking into how `isEnabled()` could be project dependent knowing that the extension is not. Turns out that you assign this variable within the lambda, which works, but is hard to follow and could break if, for example, we decide to evaluate projects in parallel. I would prefer that the `isEnabled` method takes an evaluator as a parameter.

Reply:
I used a field because, apparently, instantiating a `PluginParameterExpressionEvaluator` is expensive.

I have reworked the evaluator initialisation to make it thread-safe and dependent on the session